### ISSUE TITLE
Constructor Propagation

### ISSUE THREAD

**Issue by TonyValenti** [dotnet/roslyn#8222]
Right now if a base class has a constructor with parameters, all derived classes have to basically reimplement the constructor even if there is no functional change.  If a class does not specify _ANY_ constructors, I want it to automatically "inherit" all of it's bases constructors.  This kind of already happens today with the default constructor but I would like this extended to all constructors.

Here are a few code examples that show what I want to do.

```
//Simple base class with one parameter.
    public class Base {
        public Base(int IntParameter) {

        }
    }

//I have a derived class that can do more but takes the same constructors
//and simply passes them to the base class.  This is messy.
    public class DerivedV1 : Base {
        public void DoMore(){
        }
        public DerivedV1(int IntParameter) : base(IntParameter) {
        }
    }

//I can't do this today, but I want to.  In this case, it should get an automatic constructor:
//public DerivedV2(int IntParameter) : base(IntParameter) {
//}

    public class DerivedV2 : Base {
        public void DoMore(){
        }

    }

//Because I've specified a constructor, no constructor "propagation" occurs.
    public class DerivedV3 : Base {
        public DerivedV3(int IntParameter, string StringParameter) : base(IntParameter) {
            StringParameter.Equals(StringParameter);
        }
    }



```

In short, I want DerivedV2 to automatically "inherit" all of Base's constructors so that I don't have to litter my code with constructors that just pass values down the chain.

**Comment by HaloFour** [comment]
For reference, this is the syntax that was added in C++11:

```cpp
class Foo {
public:
    Foo(int x) { }
};

class Bar : Foo {
    using Foo::Foo; // inherit base constructors
};

...

Bar bar = Bar(123);
```

**Comment by alrz** [comment]
What if I don't want to inherit every constructor? There should be some notation to specify exactly those ctors that you want to inherit, either in the base class or the derived class.

**Comment by MgSam** [comment]
@alrz There is- exactly what you have today.

**Comment by alrz** [comment]
@MgSam Yep, so this is just a matter of tooling I guess.

**Comment by bondsbw** [comment]
@alrz If you want to remove a constructor from the set of inherited constructors, simply fall back to the existing syntax where you specify them all manually.  Including any new constructor signature erases all constructor inheritance.

Otherwise if you just want to alter one constructor from that set, you can include that constructor with the same signature and the rest would inherit.  (Not sure if that was the proposal but I'm suggesting it as a possibility.)

**Comment by alrz** [comment]
![image](https://cloud.githubusercontent.com/assets/3105979/12654630/28023b4c-c60a-11e5-9abf-988a396376c8.png)

**Comment by WrongBit** [comment]
If you **derive** class from parent, isn't it your intention to **repeat** behaviour of base + add your own? If you wanna hide base constructors, why not to introduce special keyword? Example:

    class Button {
        public Button(string ButtonText){}
    }

    class ButtonWithImage : **sealed** Button {
        public ButtonWithImage (Image img){}
    }

Here we have ButtonWithImage, which should **not** have any text (that's why we should hide constructor). But **by default** constructor should be inherited and moreover - be callable by developer at any place of derived constructor (like it done in normal languages):

    class Button {
        public Button(string ButtonText) {}
    }

    class ButtonWithImage : sealed Button {
        public ButtonWithImage (Image img){
            // some checks, actions
            inherited("simple text");// constructor is selected by signature
            // some more inits
        }
    }

This way we have full control on what and where should be done. Theory shines only on paper, but in reality we need "dirty" way, because clean theory cannot cover real life.

**Comment by HaloFour** [comment]
@WrongBit 

> If you **derive** class from parent, isn't it your intention to **repeat** behaviour of base + add your own?

Behavior, yes.  Initialization, no.  Not having constructors automatically inherited is very explicitly intentional on the part of most OOP languages as the nature of initialization is different for every class.  It does not fit in the same category as `virtual` members.

C++11 and Groovy both added explicit syntax to opt-into constructor inheritance.

>  If you wanna hide base constructors, why not to introduce special keyword?

And automatically propagate the constructors without that keyword?  That would represent a breaking change to the language, which makes it effectively a non-starter.

**Comment by HaloFour** [comment]
This seems like it would be simple enough to solve with source generators:

```cs
public class Base {
    public Base(int x) { }
    public Base(int x, int y) { }
}

[InheritConstructors]
public partial class Derived : Base { }

// generates
public partial class Derived {
    public Derived(int x) : base(x) { }
    public Derived(int x, int y) : base(x, y) { }
}
```

**Comment by gafter** [comment]
Issue moved to [dotnet/csharplang #806](https://github.com/dotnet/csharplang/issues/806) via [**ZenHub**](https://www.zenhub.com/)
